### PR TITLE
Dialog v2: Add support for returnFocusRef

### PR DESCRIPTION
**Pull request by siddharthkp**
In some rare cases, you need to return focus to a different instance of a button element than the one that triggered the Dialog (Dialog v1 supports this) 

One example of this is when you want to hide the trigger element temporarily to do some asynchronous work. Added an example in story. 

[See conversation on slack](https://github.slack.com/archives/C01L618AEP9/p1710947439592729) | [See affected code in gh/gh](https://github.com/github/github/blob/9b2f41812e3096df777bdd14105d09eb8117a7ef/ui/packages/issue-create/CreateIssueButton.tsx#L34-L51)

### Rollout strategy

- [ ] Patch release
- [x] Minor release
- [ ] Major release; if selected, include a written rollout or migration plan
- [ ] None; if selected, include a brief description as to why

### Merge checklist

- [x] Added/updated tests
- [x] Added/updated documentation
- [x] Added/updated previews (Storybook)
- [x] Changes are [SSR compatible](https://github.com/primer/react/blob/main/contributor-docs/CONTRIBUTING.md#ssr-compatibility)
- [x] Tested in Chrome
- [x] Tested in Firefox
- [x] Tested in Safari
- [x] Tested in Edge
- [ ] (GitHub staff only) Integration tests pass at github/github ([Learn more about how to run integration tests](https://github.com/github/primer-engineering/blob/main/how-we-work/testing-primer-react-pr-at-dotcom.md))

### CODE REVIEW

**Review by keithamus — COMMENTED** [review]
We might want to consider what this looks like when we migrate to native dialog, and delete the focustrap code. At that point the effective way to alter focus would be to listen for the close/cancel events and call focus on a new element. Given that's the case, and given this seems to be an edge case, why don't we document that as a pattern, rather than implementing this feature?

**Comment by siddharthkp** [comment]
> We might want to consider what this looks like when we migrate to native dialog, and delete the focustrap code. At that point the effective way to alter focus would be to listen for the close/cancel events and call focus on a new element. Given that's the case, and given this seems to be an edge case, why don't we document that as a pattern, rather than implementing this feature?

@keithamus I'd like to treat native vs synthetic dialog as an implementation. The feature here is to return focus to an element different than the one that opened the dialog, which would still be relevant (and in my opinion) a feature we can support with our Dialog wrapper over the native dialog element

**Comment by keithamus** [comment]
I guess I'm just curious about why we're enabling explicit support for rare cases? Does having a ref vs using a callback sufficiently reduce complexity for upstream users?

**Comment by siddharthkp** [comment]
> Does having a ref vs using a callback sufficiently reduce complexity for upstream users?

I think so.

If this wasn't already a feature in Dialog v1, I would have resisted it. But because it helps in transitioning from Dialog v1 to Dialog v2 seamlessly, I'm more inclined to accept this feature request. (Dialog v2 is the one we are making accessible)

File: .changeset/young-countries-confess.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+"@primer/react": patch
```
Comment (pksjce) [inline .changeset/young-countries-confess.md:2]: Just noting that its marked as `minor` on the description

Comment (siddharthkp) [reply .changeset/young-countries-confess.md]: Ah, yes you're right! Updated

**Review by pksjce — APPROVED** [review]

**Pull request merged**